docs/running-calyx/fud2/high-level-rhai.md

```diff
@@ -0,0 +1,117 @@
+
+[fud2]: ./index.md
+[fud2-design]: ./index.md#the-design-of-fud2
+
+## Example Script in High Level Rhai
+
+We'll walk through how to write a script that adds support for using the `calyx` compiler.
+
+First, we need to define some states:
+
+```rust,ignore
+export const calyx_state = state("calyx", ["futil"]);
+export const verilog_state = state("verilog", ["sv", "v"]);
+```
+
+These two lines define a `calyx` state and a `verilog` state. The `export` prefix means that these variables will be accessible to other scripts that `import "calyx"`.
+
+Now we will define an operation taking the `calyx` state to the `verilog` state. These operations define functions whose arguments are input files and return values are output values. Their bodies consist of commands that will transform those inputs to those outputs.
+
+```rust,ignore
+// define an op called "calyx_to_verilog" taking a "calyx_state" to a "verilog_state".
+defop calyx_to_verilog(c: calyx_state) >> v: verilog_state {
+    // retrieve a variable from the fud2.toml config
+    let calyx_base = config("calyx.base");
+    // retrieve a variable from the config, or a default value
+    let calyx_exe = config_or("calyx.exe", "${calyx_base}/target/debug/calyx");
+    let args = config_or("calyx.args", "");
+
+    // specify a shell command to turn a calyx file "c" into a verilog file "v"
+    shell("${calyx_exe} -l ${calyx_base} -b verilog ${args} ${c} > ${v});
+}
+```
+
+Counterintuitively, `c`, `v`, `calyx_base`, `calyx_exe`, and `args` do not contain the actual variable values. They contain identifiers which are replaced by the values at runtime. For example, `print(args)` would print a `$args` instead of the value assigned by the config. An op cannot take different code paths based on config values or different input/output file names.
+
+This example shows off nearly all of the features available for defining ops. Scripts can reuse functionality by exploiting the tools of Rhai scripting. For example, if we wanted to create a second, similar op `calyx_noverify`, we could factor the contents of `calyx_to_verilog` into a new function and call that function in both ops.
+
+```
+// a function constructing a shell command to take a calyx in_file to a verilog out_file
+// this function adds `add_args` as extra arguments to it's call to the calyx compiler
+fn calyx_cmd(in_file, out_file, add_args) {
+    let calyx_base = config("calyx.base");
+    let calyx_exe = config_or("calyx.exe", "${calyx_base}/target/debug/calyx");
+    let args = config_or("calyx.args", "");
+
+    shell("${calyx_exe} -l ${calyx_base} -b verilog ${args} ${add_args} ${in_file} > ${out_file});
+}
+
+// define an op called "calyx_to_verilog" taking a "calyx_state" to a "verilog_state".
+defop calyx_to_verilog(c: calyx_state) >> v: verilog_state {
+    calyx_cmd(c, v, "");
+}
+
+// define an op called "calyx_noverify" taking a "calyx_state" to a "verilog_state".
+defop calyx_to_verilog(c: calyx_state) >> v: verilog_state {
```

Comment:
Seems like a couple of things are out of date on these lines: the op should be named `calyx_noverify` instead of `calyx_to_verilog`, and the output state (both in the comment and in the actual `defop` line) should be something else?

Reply:
Thanks for catching this!